numba/core/types/__init__.py

```diff
@@ -40,151 +40,302 @@
+    ushort = np_ushort = _make_unsigned(np.short)
+    int_ = np_int_ = _make_signed(np.int_)
+    uint = np_uint = _make_unsigned(np.int_)
+    intc = np_intc = _make_signed(np.intc) # C-compat int
+    uintc = np_uintc = _make_unsigned(np.uintc) # C-compat uint
+    long_ = np_long = _make_signed(np.int_)  # C-compat long
+    ulong = np_ulong = _make_unsigned(np.int_)  # C-compat ulong
+    longlong = np_longlong = _make_signed(np.longlong)
+    ulonglong = np_ulonglong = _make_unsigned(np.longlong)
+
+    all_str = '''
+    int8
+    int16
+    int32
+    int64
+    uint8
+    uint16
+    uint32
+    uint64
+    intp
+    uintp
+    intc
+    uintc
+    ssize_t
+    size_t
+    boolean
+    float32
+    float64
+    complex64
+    complex128
+    bool_
+    byte
+    char
+    uchar
+    short
+    ushort
+    int_
+    uint
+    long_
+    ulong
+    longlong
+    ulonglong
+    float_
+    double
+    void
+    none
+    b1
+    i1
+    i2
+    i4
+    i8
+    u1
+    u2
+    u4
+    u8
+    f4
+    f8
+    c8
+    c16
+    optional
+    ffi_forced_object
+    ffi
+    deferred_type
+    '''
+else:
+    from .new_scalars import *
+    ### Machine Datatypes ###
+    c_bool = MachineBoolean('c_bool')
+    c_byte = c_int8 = MachineInteger('c_int8')
+    c_int16 = MachineInteger('c_int16')
+    c_int32 = MachineInteger('c_int32')
+    c_int64 = MachineInteger('c_int64')
+    c_uint8 = MachineInteger('c_uint8')
+    c_uint16 = MachineInteger('c_uint16')
+    c_uint32 = MachineInteger('c_uint32')
+    c_uint64 = MachineInteger('c_uint64')
+
+    c_intp = c_int32 if utils.MACHINE_BITS == 32 else c_int64
+    c_uintp = c_uint32 if utils.MACHINE_BITS == 32 else c_uint64
+
+    # Machine Floats
+    c_float16 = MachineFloat('c_float16')
+    c_float32 = MachineFloat('c_float32')
+    c_float64 = MachineFloat('c_float64')
+
+    # Machine Complex
+    c_complex64 = MachineComplex('c_complex64', c_float32)
+    c_complex128 = MachineComplex('c_complex128', c_float64)
+
+    c_signed_domain = frozenset([c_int8, c_int16, c_int32, c_int64])
+    c_unsigned_domain = frozenset([c_uint8, c_uint16, c_uint32, c_uint64])
+    c_integer_domain = c_signed_domain | c_unsigned_domain
+    c_real_domain = frozenset([c_float32, c_float64])
+    c_complex_domain = frozenset([c_complex64, c_complex128])
+    c_number_domain = c_real_domain | c_integer_domain | c_complex_domain
+
+    ### Python Datatypes ###
+    # Python Integers
+    py_bool = PythonBoolean('py_bool')
+    py_int64 = PythonInteger('py_int64')
+
+    py_intp = py_int64
+
+    # Python Float
+    py_float64 = PythonFloat('py_float64')
+
+    # Python Complex
+    py_complex128 = PythonComplex('py_complex128', py_float64)
+
+    py_signed_domain = frozenset([py_int64])
+    py_integer_domain = py_signed_domain
+    py_real_domain = frozenset([py_float64])
+    py_complex_domain = frozenset([py_complex128])
+    py_number_domain = py_real_domain | py_integer_domain | py_complex_domain
```

Comment:
We may want to let them be. Otherwise to prevent changes in logic common across the old and new type system we'd need to redefine the domains in lots of places individually.